Homepage stats auto

akmiller01 opened the pull request:
Alternative solution to homepage stat updater that does it all in a management command. Will make cronjob easier.

akmiller01 commented:
Pair PR on deployment: https://github.com/IATI/deployment/pull/36

akmiller01 commented:
Might want to merge https://github.com/IATI/preview-website/pull/129 first

dalepotter commented on iati/home/management/commands/updatestatistics.py:
```diff
@@ -0,0 +1,39 @@
+import requests
```
I would explicitly add `requests` to `requirements.txt` as it is an explicit dependency.

dalepotter replied:
Resolved in 8bbaa86

dalepotter commented on iati/home/management/commands/updatestatistics.py:
```diff
@@ -0,0 +1,39 @@
+    help = 'Update statistics'
+
+    def handle(self, *args, **options):
+        """The default function Django BaseCommand needs to run."""
```
I would definitely break this up into a couple further functions...

- lines 18-27 into something like `get_total_num_activities` with docstring to state it queries the Registry, loops over datasets and totals reported number of activities.

- lines 29-36 into something like `get_total_num_publishers` with similar docstring.

dalepotter replied:
Ideally we'd unit test these functions, using mocked data from the Registry - we've had some success with the [responses](https://github.com/getsentry/responses) module. There's an [example in action here](https://github.com/IATI/iati.fetch/blob/fetch-dataset/iati/fetch/tests/test_dataset.py#L86-L120) (which tests [this function](
https://github.com/IATI/iati.fetch/blob/fetch-dataset/iati/fetch/__init__.py#L104-L123)).

(This functionality was what the `fetch` part of pyIATI was intended to be about - as started in this code. It never got completed as we focused on other parts of pyIATI, but those links should provide a good basis for testing).

dalepotter replied:
Refactoring resolved in ce3c1bc Testing resolved in that and further commits.

akmiller01 commented:
Hey @Dale I think I addressed all of your comments:

- Refactored command into two functions with two URL variables (easier import into unit test)
- Unit testing using responses and mock JSON downloaded from the registry
- Added explicit references to `requests` version and `responses` in `requirements.txt` and `requirements_dev.txt` respectively.

dalepotter commented on iati/tests/data/activities.json:
```diff
@@ -0,0 +1 @@
+{"help": "https://iatiregistry.org/api/3/action/help_show?name=package_search", "success": true, "result": {"count": 5699, "sort": "title_string asc", "facets": {"extras_activity_count": {"693": 1, "4029": 1, "3551": 1, "1940": 1, "3253": 1, "3558": 1, "818": 2, "0": 4, "346": 1, "347": 1, "992": 1, "348": 1, "349": 3, "816": 1, "2911": 1, "718": 1, "1793": 1, "715": 1, "426": 1, "1495": 1, "6842 ...
```
Great that we've added test data files in their own folder. 

As a note, typically these files tend to include a minimal subset of expected data (and be pretty formatted) so that you can see the expected structure of the file.

dalepotter commented on iati/tests/events_functional_tests.py:
```diff
@@ -215,6 +215,44 @@ def test_event_type_filter(self, admin_browser):
         admin_browser.click_link_by_text(TEST_CATEGORY)
         assert admin_browser.is_text_present("Show all events")
 
+    # Uncomment once we merge editable-header-image
```
Minor point - typically things such as this tend to be prepended with `TODO` - that way we can easily search through the repo at any point to find leftover tasks.

dalepotter commented on iati/tests/management_unit_tests.py:
```diff
@@ -0,0 +1,21 @@
+
+@pytest.mark.django_db
+@responses.activate
+def test_update_statistics():
```
This is great! Would suggest a test for the exceptions that we're hoping get raised too :-)

dalepotter commented on iati/tests/events_functional_tests.py:
```diff
@@ -7,8 +7,7 @@
 from django.utils.text import slugify
-from base_functional_tests import find_and_click_add_button, find_and_click_toggle_button, fill_content_editor_block
-
+from base_functional_tests import click_obscured, find_and_click_add_button, find_and_click_toggle_button, fill_content_editor_block, TEST_DATA_DIR
```
Bit confused about how these changes have made it into this PR??

dalepotter commented:
Approved - would still recommend making the test data more simplified (otherwise debugging failures related to this could be a bit bewildering).

Could anyone explain why the changes for events tests have made it into this branch? I'm confused given I'd expect those changes to already be in the base branch...
